week2 level1 과제

WonJoongLee opened the pull request:
- level1 과제를 마쳤습니다.
- 1차 세미나 때 부족했던 부분(registerForActivityResult 미사용, 네이밍 컨벤션 없음, 의미 없는 변수명 등)을 고쳤습니다.

WonJoongLee committed:
(#3) 리사이클러뷰 위한 xml 작업

- activity_home.xml에 리사이클러 뷰 추가
- item_repository.xml 추가

WonJoongLee committed:
(#3) 폴더 정리

- ui, data 폴더 생성해서 파일들을 옮겼음

WonJoongLee committed:
(#3) Level1 리사이클러뷰 완료

- Level 1에서 요구하는 리사이클러 뷰 완료
- 설명이 길 경우 뒤에 "..." 추가

WonJoongLee committed:
(#3) 리사이클러뷰 data 추가 방식 변경

- for문으로 반복적인 코드를 줄임.

WonJoongLee committed:
(#3) Fragment Activity 추가

- Home Activity에서 More을 클릭하면 Fragment를 가진 Activity로 넘어감.
- 2차 세미나 level1 과제는 마무리 되었음.

WonJoongLee committed:
(#4) registerForActivityResult 구현

WonJoongLee committed:
(#4) xml에 네이밍 컨벤션 적용

WonJoongLee committed:
(#4) 조건 함수 추가, upper snake case 적용, 의미 있는 변수명, 함수명 적용

WonJoongLee committed:
(#3) 함수로 깔끔한 코드 생성.

- 기존에 onCreate()에서 리사이클러 뷰 세팅을 처음부터 끝까지 다 해줬는데, `private fun setRepoRv` 함수를 생성하여 코드를 깔끔하게 처리했음.

WonJoongLee committed:
(#3) 저를읽어주세요.md 추가

WonJoongLee committed:
저를읽어주세요

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/adapters/RepositoryAdapter.kt:
```diff
@@ -0,0 +1,37 @@
+
+class RepositoryAdapter : RecyclerView.Adapter<RepositoryAdapter.RepositoryViewHolder>() {
+
+    var data = mutableListOf<RepositoryInfo>()
```
어제 내가 말한 부분이 이 부분인데, 리스트를 만들 때 유동적으로 변할 수 있고(Mutable), 변수 자체도 바뀔 수 있도록 설정을 한 것은 실수를 초래할 수 있을 것 같아. 

만약에 내가 서버 통신을 하고 data에 새로운 리스트를 그냥 repositoryAdapter.data = newList 이런식으로 대입을 해준다면 어떤 문제가 발생할까? 이 부분을 다시 생각해봤으면 좋겠어

WonJoongLee replied:
1. 서버 통신하면 `notifyDataSetChanged()`만 해주면 문제 없을 것 같았는데, [공식문서](https://developer.android.com/reference/android/widget/BaseAdapter#notifyDataSetChanged())에 `any View reflecting the data set should refresh itself.` 이렇게 나와있는 걸 보면 계속 리사이클러뷰를 새로고침해서 메모리가 많이 사용되는 문제가 있을 것 같기도 한데 잘은 모르겠어.
2. newList로 추가를 하면 처음부터 아이템이 계속 추가가 되나? 예를 들면 1\~10까지의 item이 있는데 11\~13의 item이 들어오면 기존 뷰에 있던 1\~10에 이어서 뒤에 1\~13의 item이 들어오는건가?

정확히는 잘 모르겠네 ㅋㅋ 사실 바뀌지 않을 값이면 val로 하는 것이 맞기 때문에 저 부분은 고쳐하는데, 뒤에서 말한 새로운 newList를 대입하면 어떤 문제가 초래할 지까지는 잘 모르겠어

l2hyunwoo replied:
그럼 이렇게 생각을 바꿔보자
- 너가 한건 newList를 대입한걸까? newList를 참조한걸까?

WonJoongLee replied:
참조인 것 같아. 아래 코드로 참조(shallow copy)와 대입(deep copy)의 차이점을 확인했어.
```
fun main() {
    data class Person(val age : Int, val name : String)
    val origin = ArrayList<Person>()
    origin.add(Person(7, "obama"))
    origin.add(Person(3, "trump"))
    var newData = ArrayList<Person>()
    newData = origin // 이렇게 대입하면 shallow copy
    //newData.addAll(origin) // 이렇게 값들을 넣어주면 deep copy
    newData.removeAt(1)
    println(origin)
}
```
그러니까 얕은 복사여서 현재 리사이클러 뷰가 알고 있는 data와, 서버에서 변경된 data로 인해 실제 데이터가 서로가 다른거야?

l2hyunwoo replied:
얕은 복사여서 두 데이터가 참조를 하게되는 주소값이 같아지고 그러면 너가 실수를 하게 된다면 데이터가 잘못될 가능성도 있지 않을까? 코딩할 때 중요한 것중 하나는 실수를 발생시킬 가능성을 원천 차단하는 거라고 생각해.

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/adapters/RepositoryAdapter.kt:
```diff
@@ -0,0 +1,37 @@
+
+    var data = mutableListOf<RepositoryInfo>()
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RepositoryViewHolder {
+        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_repository, parent, false)
+        return RepositoryViewHolder(view)
+    }
```
View(Data) Binding도 배웠는데 이걸 활용해서 작성하는 건 어떨까?

WonJoongLee replied:
확인! 수정해볼게

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    private fun repoDataInput(repoList: MutableList<RepositoryInfo>): MutableList<RepositoryInfo> {
+        for (i in 1..10) {
+            repoList.add(RepositoryInfo("레포지토리 이름".plus(i), "레포지토리 설명".plus(i), "레포지토리 언어".plus(i)))
+        }
+        //Repository 이름이나 설명이 긴 경우 처리를 보여주기 위해 임시로 추가
+        repoList.add(
+            RepositoryInfo(
+                "레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름",
+                "레포지토리 설명레포지토리 설명레포지토리 설명레포지토리 설명레포지토리 설명",
+                "레포지토리 언어"
+            )
+        )
+        return repoList
+    }
```
매개변수로 들어온 repoList에 add를 하게 되면 setRepoRv에서 선언한 repoList의 값은 그대로 유지될까? 코틀린에서 대입은 주소값을 넘겨주는 거라고 생각하는게 더 좋아. 직접 로그값을 찍어봐서 어떻게 달라지는 지 확인해보고 리팩토링하면 좋을 것 같다

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -0,0 +1,94 @@
+        val repoRecyclerView = findViewById<RecyclerView>(R.id.rv_repository)
+        repoList = repoDataInput(repoList)
+        repoAdapter.data = repoList
+        repoRecyclerView.layoutManager =
+            LinearLayoutManager(applicationContext, LinearLayoutManager.VERTICAL, false)
```
음....Activity(뷰)에서 사용하는건데 applicationContext를 사용하면 어떤 일이 발생할까?

WonJoongLee replied:
그렇네. activity에서 사용하는 거니까 `basecontext`를 사용해야 할 것 같아. Activity 내에서 일어나는 일이니까 `baseContext`를 사용해야 한다는 것은 알겠는데, applicationContext를 사용하면 어떤 문제가 생기지??

l2hyunwoo replied:
- 굳이 baseContext보다 그냥 this를 사용해도 되지 않을까?(그리고 나는 이런거 xml에서 관리시킴)
- applicationContext의 생명주기가 길까? 아니면 View의 생명주기가 길까?

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -0,0 +1,94 @@
+    private fun setRepoRv() {
+        var repoList = mutableListOf<RepositoryInfo>()
+        val repoAdapter = RepositoryAdapter()
+        val repoRecyclerView = findViewById<RecyclerView>(R.id.rv_repository)
```
binding 객체가 있는데 굳이 findViewById를 활용할 이유가 있을까?

WonJoongLee replied:
확인! 수정할게요

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/SignInActivity.kt:
```diff
@@ -36,56 +47,44 @@ class SignInActivity : AppCompatActivity() {
-            val intent = Intent(this, SignUpActivity::class.java)
-            startActivityForResult(intent, 0)
-        }
+    private fun isIdPwETEmpty() : Boolean{
+        return binding.etGithubId.text.isNullOrBlank() || binding.etGithubPw.text.isNullOrBlank()
```
```suggestion
    private fun isIdPwETEmpty() = binding.etGithubId.text.isNullOrBlank() || binding.etGithubPw.text.isNullOrBlank()
```

한줄로 작성할 수 있는 함수는 그냥 "="을 적어도 좋습니당

l2hyunwoo reviewed (changes requested):
수고했어! 리사이클러 뷰에서 개발자의 실수가 발생할 경우 큰 에러가 날 수 있는 부분이 있는 것 같아 그 부분을 다시 수정하면 좋을 것 같아.

SeojinSeojin commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -23,31 +22,42 @@ class HomeActivity : AppCompatActivity() {
         repoAdapter.data = repoList
-        repoRecyclerView.layoutManager = LinearLayoutManager(applicationContext, LinearLayoutManager.VERTICAL, false)
+        repoRecyclerView.layoutManager =
+            LinearLayoutManager(applicationContext, LinearLayoutManager.VERTICAL, false)
```
나는 이부분 xml에서 했다!

SeojinSeojin commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -15,13 +15,13 @@ import org.sopt.androidseminar.databinding.ActivityHomeBinding
 
     private lateinit var binding : ActivityHomeBinding
-    private val activityName = "HomeActivity"
+    private val ACTIVITYNAME = "HomeActivity"
```
이친구는 companion object로 보내라고 하네!
```kotlin
companion object {
    private val ACTIVITYNAME = "HomeActivity"
}
```

SeojinSeojin replied:
그리고 upper snake case로 쓴다면 `ACTIVITY_NAME`도 괜찮을 것 같아!
다만 지난주랑 겹치는 부분이니까 크게 신경쓰진 않아도 될 것 가타ㅏ

WonJoongLee replied:
아 그렇구나 언더바 추가해야겠다.
companion object도 어떤건지 공부해서 반영하겠슴다!!

SeojinSeojin commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/FragmentActivity.kt:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    private fun connectFragment() {
+        val tempFragment = TempFragment()
```
계속 이름 갖고 그러는데! 이 프라그먼트 친구도 기능 명시된걸로 바꾼다면 좋겠따!!

WonJoongLee replied:
오키 temp는 넘했네

SeojinSeojin commented on Second_Seminar/app/src/main/res/layout/activity_fragment.xml:
```diff
@@ -0,0 +1,33 @@
+
+    <androidx.fragment.app.FragmentContainerView
+        android:id="@+id/fragment_temp"
+        android:layout_width="match_parent"
```
이 친구도 `0dp`여도 괜찮을 거 같아!!

SeojinSeojin commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -34,16 +34,11 @@ class HomeActivity : AppCompatActivity() {
+            repoList.add(RepositoryInfo("레포지토리 이름".plus(i), "레포지토리 설명".plus(i), "레포지토리 언어".plus(i)))
+        }
+        //Repository 이름이나 설명이 긴 경우 처리를 보여주기 위해 임시로 추가
+        repoList.add(RepositoryInfo("레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름", "레포지토리 설명레포지토리 설명레포지토리 설명레포지토리 설명레포지토리 설명", "레포지토리 언어"))
```
오우 신선한데.. 나도 이런거 하나 넣어야겠다!

SeojinSeojin reviewed (commented):
수고하셨습니다! 👍👍

WonJoongLee committed:
(#3) pull request 내용 중 일부 반영

- 네이밍 컨벤션 일부 수정
- Application Context->Activity Context로 수정
- findviewbyid->binding 객체로 접근
- 함수 중괄호 쓰지 않고 =을 통해 구현

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/adapters/RepositoryAdapter.kt:
```diff
@@ -0,0 +1,37 @@
+    override fun getItemCount(): Int = data.size
+
+    class RepositoryViewHolder(itemView : View) : RecyclerView.ViewHolder(itemView){
+        private val repoName = itemView.findViewById<TextView>(R.id.tv_repository_name)
+        private val repoInfo = itemView.findViewById<TextView>(R.id.tv_repository_detail)
+        private val repoLang = itemView.findViewById<TextView>(R.id.tv_repository_language)
+        fun bind(repositoryInfo: RepositoryInfo){
+            repoName.text = repositoryInfo.repoName
+            repoInfo.text = repositoryInfo.repoInfo
+            repoLang.text = repositoryInfo.repoLang
+        }
```
여기서 조금만 더 나아가면 DataBinding으로 더 깔끔하게 정리할 수 있거든? 한번 해보실?

mdb1217 replied:
ㄹㅇㄹㅇ DataBinding 한 번 잡숴봐... 저런 코드 3줄 안 써도 다 해결됌 ㄹㅇ

WonJoongLee replied:
오키 한 30초 전에 viewholder에서 viewbinding되는지 알았는데 Data Binding으로도 해보겠슴다!!

l2hyunwoo reviewed (commented):
추가적인 의견 더 달아봤어~

mdb1217 commented on Second_Seminar/app/build.gradle:
```diff
@@ -9,7 +9,7 @@ android {
     defaultConfig {
         applicationId "org.sopt.androidseminar"
-        minSdkVersion 28
+        minSdkVersion 23
```
개인적으로 궁금한것..! SdkVersion을 23으로 바꾼 이유가 뭐야?

WonJoongLee replied:
더 많은 기종에서 앱 사용할 수 있으면 좋을 것 같아서 낮췄어. 배포할 게 아니어서 상관 없을 것 같긴 한데, 평소에 21, 23에 두고 개발하다 보니까 이번에도 저렇게 바꿨어ㅋㅋ

mdb1217 commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/adapters/RepositoryAdapter.kt:
```diff
@@ -0,0 +1,37 @@
+    override fun getItemCount(): Int = data.size
+
+    class RepositoryViewHolder(itemView : View) : RecyclerView.ViewHolder(itemView){
+        private val repoName = itemView.findViewById<TextView>(R.id.tv_repository_name)
+        private val repoInfo = itemView.findViewById<TextView>(R.id.tv_repository_detail)
+        private val repoLang = itemView.findViewById<TextView>(R.id.tv_repository_language)
```
findViewById의 문제점은 이미 원중오빠도 잘 알거라고 생각하지만..
- Null safety 하지못함
- 사용하기에 꽤나 무거움
- 코드가 더러워짐(...)

위와 같은 이유로 현우오빠도 위에서 View(Data)Binding을 적용하라고 한 것 같다..!

WonJoongLee replied:
viewholder에서도 viewbinding이나 databinding이 되는지 몰랐네 공부해서 수정해야겠다

mdb1217 commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    private fun repoDataInput(repoList: MutableList<RepositoryInfo>): MutableList<RepositoryInfo> {
+        for (i in 1..10) {
+            repoList.add(RepositoryInfo("레포지토리 이름".plus(i), "레포지토리 설명".plus(i), "레포지토리 언어".plus(i)))
+        }
+        //Repository 이름이나 설명이 긴 경우 처리를 보여주기 위해 임시로 추가
+        repoList.add(
+            RepositoryInfo(
+                "레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름레포지토리 이름",
+                "레포지토리 설명레포지토리 설명레포지토리 설명레포지토리 설명레포지토리 설명",
+                "레포지토리 언어"
+            )
+        )
+        return repoList
```
굳이 전역 변수로 선언 된 repoList를 다시 한 번 함수의 매개변수로 넣어 줄 필요가 있을까? 그냥 function 내부에서 repoList를 바로 참조해서 add 작업을 해도 상관없을 것 같은데 혹시 이런 식으로 작업한 이유가 있을까?

WonJoongLee replied:
@mdb1217 repoList가 setRepoRv() 내부에서 선언됐는데 이러면 지역변수 아니야?!?!

WonJoongLee committed:
(#3) 코드리뷰 반영

- 리사이클러뷰 data binding으로 수정
- repoDataInput함수에서 반환값 제거
- 리사이클러뷰 LinearLayoutManager 수정
- findViewById 제거
- 한 줄로 작성할 수 있는 함수 수정
- companion object 추가

WonJoongLee committed:
(#3) GridLayout 추가

- TempFragment를 깃헙 팔로워들로 변경함.
- 팔로워들을 GridLayout으로 나열함.

l2hyunwoo commented on First_Seminar/app/src/main/java/org/sopt/androidseminar/SignInActivity.kt:
```diff
@@ -5,13 +5,26 @@ import androidx.appcompat.app.AppCompatActivity
+        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
+            if (it.resultCode == RESULT_OK) {
+                val data = it.data
+                data?.let {
```
let 사용하는거 아주 좋습니다!

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/adapters/FollowerAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+import org.sopt.androidseminar.data.FollowerInfo
+import org.sopt.androidseminar.databinding.ItemFollowerBinding
+
+class FollowerAdapter(private val data: List<FollowerInfo>) :
```
- 우선 data를 private val로 받아오는 방식은 난 좋다고 생각함
- 근데 이런 방식으로 어댑터 코드를 작성하면 데이터 갱신을 어떻게 할 수 있을까?

WonJoongLee replied:
아 그러면 클래스 내부로 `private val` 부분을 빼줘야 겠군욥

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/adapters/FollowerAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+            binding.apply {
+                tvFollowerName.text = followerInfo.followerName
+                tvFollowerGithubId.text = followerInfo.followerGithubId
+            }
```
위에서 Databinding을 사용했다면 다른 방식도 사용가능해

WonJoongLee replied:
현우야 이거 금방 Databinding으로 반영해서 푸쉬다시 날렸어

l2hyunwoo replied:
@WonJoongLee 확인하고 코멘트 남김

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/adapters/FollowerAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+        holder.bind(data[position])
+    }
+
+    override fun getItemCount(): Int = data.size
```
```suggestion
    override fun getItemCount() = data.size
```

'='을 사용하면 반환형 표기 안해도 됨

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/FollowerFragment.kt:
```diff
@@ -0,0 +1,51 @@
+
+class FollowerFragment : Fragment() {
+
+    private lateinit var binding: FragmentFollowerBinding
```
지난번 세미나에서 진수가 말해줬듯이 binding의 생명주기를 참고해서 binding객체를 만들면 더 좋을 것 같아

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/FollowerFragment.kt:
```diff
@@ -0,0 +1,51 @@
+        return binding.root
+    }
+
+    private fun setFollowerRv() {
```
함수 이름으로 [되도록이면 축약어를 사용하지 않으면](https://ilovewoojin.wordpress.com/2015/05/15/%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%A8program-%EB%84%A4%EC%9D%B4%EB%B0%8Dnaming-%EA%B7%9C%EC%B9%99/) 좋을 것 같아

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/FollowerFragment.kt:
```diff
@@ -0,0 +1,51 @@
+        followerList.add(FollowerInfo("오바마", "obama"))
+        for (i in 0..10) {
+            followerList.add(FollowerInfo("트럼프", "trump"))
+        }
```
```suggestion
        (0..10).forEach { followerList.add(FollowerInfo("트럼프", "trump")) }
```

lambda를 사용하면 조금 더 가독성이 높아질 수 있어

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -0,0 +1,94 @@
+    private fun repoDataInput(repoList: MutableList<RepositoryInfo>) {
+        for (i in 1..10) {
+            repoList.add(RepositoryInfo("레포지토리 이름".plus(i), "레포지토리 설명".plus(i), "레포지토리 언어".plus(i)))
+        }
```
```suggestion
    (1..10).forEach {  repoList.add(RepositoryInfo("레포지토리 이름".plus(it), "레포지토리 설명".plus(it), "레포지토리 언어".plus(it)) }
```

forEach/map/forEachIndexed 등 다양한 함수형 구문을 활용해서 간결하게 루프를 바꾸면 가독성이 올라갑니당

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -0,0 +1,94 @@
+        binding = ActivityHomeBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        Log.d(ACTIVITY_NAME, "onCreate")
+
+        setRepoRv()
+
+        moreButtonClickedEvent()
```
불필요한 Spacing은 굳이 안해도 될 것 같아
```suggestion
        setRepoRv()
        moreButtonClickedEvent()
```

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -0,0 +1,94 @@
+        repoDataInput(repoList)
+
+        val repoAdapter = RepositoryAdapter(repoList)
+        val repoRecyclerView = binding.rvRepository
```
음 View를 변수로 받아와야하는 이유를 내가 생각하지 못해서 그러는데 알려줄 수 있을까?

WonJoongLee replied:
어댑터 설정하려고 가져왔어!!
지금 태그해준 문장 바로 아래에 
``` 
repoRecyclerView.adapter = repoAdapter
```
이렇게 어댑터 설정해주는데...

WonJoongLee replied:
어댑터 설정하려고 가져왔어!!
지금 태그해준 문장 바로 아래에 
``` 
repoRecyclerView.adapter = repoAdapter
```
이렇게 어댑터 설정해주는데... 이게 정석이 아닌가!?

WonJoongLee committed:
(#3) RecyclerView item에 Databinding 적용

l2hyunwoo commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    private fun repoDataInput(repoList: MutableList<RepositoryInfo>) {
```
- 오 함수 내용은 매우 굳! 👍 
- 데이터를 init하는 거라면 ``init~`` ``set~`` 선에서 네이밍 했을 것 같아

l2hyunwoo reviewed (commented):
역시 원중, 뭐하나 부족하지 않고 잘 짜여진 것 같아. 내가 느낀 점들을 피드백했으니 확인 부탁해 너무 수고했다!

l2hyunwoo commented on Second_Seminar/app/src/main/res/layout/item_follower.xml:
```diff
@@ -0,0 +1,63 @@
+
+    <data>
+
+        <variable
+            name="name"
+            type="String" />
+
+        <variable
+            name="githubNickName"
+            type="String" />
```
- 👍 좋습니다
- 하지만 variable에서 type을 반드시 primitive로 연결시킬 필요는 없어, 이미 Data class에 한 묶음으로 정의하지 않았나?

l2hyunwoo reviewed (commented):
Data Binding 적용한거 추가적으로 comment 달았어

WonJoongLee committed:
(#6) Recyclerview에 광고 item 추가

- GridLayout은 구현되었음.
- Level 2-2 Recycler View에 여러 view 보여주기 과제 중

WonJoongLee committed:
(#6) ViewType을 FollowerInfo에 추가

- Companion Object로 추가하였음.

WonJoongLee committed:
(#6) Multi ViewType을 통해 광고 추가

- GridLayout 중간에 광고가 나오도록 설정 완료

WonJoongLee committed:
(#3) for문 -> forEach로 수정

WonJoongLee committed:
(#3) 불필요한 space 제거

WonJoongLee committed:
(#3) 모호한 함수명 수정

WonJoongLee committed:
(#3) Data binding에서 data class로 type 설정

- Follower Adapter에서 primitive type을 FollowerInfo data class type으로 연결하였음.
- Repository Adapter에서 primitive type을 RepositoryInfo data class type으로 연결하였음.

WonJoongLee committed:
(#3) 불필요한 반환형 제거

WonJoongLee committed:
(#3) 불필요한 반환형 제거

WonJoongLee committed:
(#3) 의미있는 변수명으로 변경

- FragmentActivity.kt에서 tempFragment -> followerFragment

WonJoongLee committed:
(#3) 저를읽어주세요

todayiswindy commented:
수고했어! 최종 수정하기 이전 코드들도 쭉 보고 있었는데 그동안 피드백 받았던 거 다 수정해온 게 대단하다고 생각했어 다른 친구들이 리뷰를 잘 남겨줘서 내가 따로 더 말해줄 건 없는 것 같습니다 많이 배우고 갑니다 :D

todayiswindy commented:
수고했어! 최종 수정하기 이전 코드들도 쭉 보고 있었는데 그동안 피드백 받았던 거 다 수정해온 게 대단하다고 생각했어 다른 친구들이 리뷰를 잘 남겨줘서 내가 따로 더 말해줄 건 없는 것 같습니다 많이 배우고 갑니다 :D

jinsu4755 commented on Second_Seminar/app/src/main/java/org/sopt/androidseminar/data/FollowerInfo.kt:
```diff
@@ -0,0 +1,8 @@
+package org.sopt.androidseminar.data
+
+data class FollowerInfo(val followerName: String, val followerGithubId: String, val viewType : Int){
+    companion object{
+        const val NORMAL_CONTENT = 0
+        const val AD_CONTENT = 1
+    }
+}
```
이건 약간 물어보고 싶은건데 해당 팔로우 인포는 결국 앱안에서 쓰는 데이터의 역할을 가지게 될 것 같은데
뷰의 타입은 데이터일 수 있지만 이것은 뷰의 데이터이지 팔로우 정보의 데이터라고 하긴 힘들 것같다는 생각이 들어서 해당 부분은 물어보고 싶어~

데이터들을 받아서 이 데이터는 어떤 뷰홀더가 그려야할지 구분하는 역할은 어떤 친구가 해야할까?

WonJoongLee replied:
1. 단순히 data class여서 data에 넣었는데, view에 포함되는 게 맞는 것 같아ㅋㅋ view를 보여주기 위한 data class니까. 근데 다시 생각해보면 view를 보여줄 때 data class를 활용하는 것은 맞지만, 팔로잉 값들을 받을 때 data class로 사용되니까 data에 포함되어야 할 것 같기도 해. 좀 헷갈리네,,

2. 데이터들을 넣을 때 마지막 인자 값 viewType으로 어떤 view인지 파악해서 각각 맞는 viewHolder를 찾아가게 설정했어. 그래서 구분하는 역할을 viewType이 하기는 했는데, 혹시 다른 방법으로 해야하나..ㅎㅎ

jinsu4755 replied:
아하 일단 질문의 의도는 뷰타입에대한 상수의 위치는 어디일까? 와 뷰타입이란 값을 데이터 클레스에서 가져야하는가? 인데
형이 어떻게 쓴지는 이해했어!
뷰타입에 대한 상수! 뷰 타입이 0인지 1일지는 뷰에대한 정보가 될 수 있지!
그러나 그 0과 1을 포장한 값들마져 데이터클래스 안에 있어야 할까?

난 이부분이 무언가 공장을 예를 들자면
앞쪽 라인(엑티)에서 포장라인(어뎁터)으로 제품(데이터1개)들이 나오고있는데 불량인지 정상인지 검사하는데 그 기준이 제품안에 있는느낌?
제품이 자기 자신이 불량이란걸 아는 느낌이 된다고 생각해!

그리고 두번째!
지금은 우리가 담아주지만 다른곳에서 온다면 우리가 다시 매핑해줘야한다는 문제가 생기지 않을까?

각설하고 핵심만 뽑자면
사실 세미나에서 해당 내용을 다루지 않기 때문에 어뎁터는 어디선가 자료를 받아 "정리한다" 라고 이야기 했지만
어뎁터는 뷰와 데이터 사이의 연결방법을 "변경" 시킨다라고도 써둔거 같아
변경은 뷰홀더가 만들때 뷰타입으로 변경한다면

어뎁터는 자료를 어떻게 정리할까?
어뎁터가 사실은 자료에 어떤 뷰타입인지 붙여주는 역할도 있는게 아닐까?
만약 어뎁터가 이런 자료 정리, 어떤데이터가 어떤 뷰타입인지 붙여주는 역할도 하도록 설계되었다면
우리가 상속받는 리사이클러뷰의 어뎁터에 무언가 있지 않을까?!

jinsu4755 reviewed (commented):
너무 잘해줘서 고맙고 2차 과제 고생많았어~~
